File: answers/other/answers/lintcode/N-Queens.cpp
class Solution {
public:
    /**
     * Get all distinct N-Queen solutions
     * @param n: The number of queens
     * @return: All distinct solutions
     * For example, A string '...Q' shows a queen on forth position
     */
    vector<vector<string> > solveNQueens(int n) {
        // write your code here
		nqueens_list.clear();
		vector<string> nqueens(n, string(n, '.'));
		solveNQueens(nqueens, 0);
		return nqueens_list;
    }
	void solveNQueens(vector<string> &nqueens, int i) {
		if (i >= nqueens.size()) {
			nqueens_list.push_back(nqueens);
			return;
		}
		for (size_t j = 0; j < nqueens[i].size(); ++ j) {
			if ('.' == nqueens[i][j]) {
				bool flag = true;
				for (size_t k = 0; k < nqueens.size(); ++ k) {
					if ('Q' == nqueens[k][j]) {
						flag = false;
						break;
					}
				}
				if (flag) {
					for (size_t m = i - 1, n = j - 1; m < nqueens.size() && n < nqueens[m].size(); -- m, -- n) {
						if ('Q' == nqueens[m][n]) {
							flag = false;
							break;
						}
					}
				}
				if (flag) {
					for (size_t m = i - 1, n = j + 1; m < nqueens.size() && n < nqueens[m].size(); -- m, ++ n) {
						if ('Q' == nqueens[m][n]) {
							flag = false;
							break;
						}
					}
				}
				if (flag) {
					nqueens[i][j] = 'Q';
					solveNQueens(nqueens, i + 1);
					nqueens[i][j] = '.';
				}
			}
		}
	}
	vector<vector<string> > nqueens_list;
};

File: answers/leetcode/N-Queens/N-Queens.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> queens = new LinkedList<>();
        this.dfs(new LinkedList<>(), n, queens);
        return queens;
    }

    private void dfs(List<Integer> positions, int n, List<List<String>> queens) {
        if (positions.size() == n) {
            List<String> queen = new LinkedList<>();
            for (int i = 0; i < n; i++) {
                StringBuilder sb = new StringBuilder();
                for (int j = 0; j < n; j++) {
                    sb.append(j == positions.get(i) ? 'Q' : '.');
                }
                queen.add(sb.toString());
            }
            queens.add(queen);
        }

        for (int i = 0; i < n; i++) {
            boolean flag = true;
            for (int j = 0; j < positions.size(); j++) {
                if (i == positions.get(j) || i + positions.size() == j + positions.get(j) || i - positions.size() == positions.get(j) - j) {
                    flag = false;
                    break;
                }
            }
            if (flag) {
                positions.add(i);
                this.dfs(positions, n, queens);
                positions.remove(positions.size() - 1);
            }
        }
    }
}